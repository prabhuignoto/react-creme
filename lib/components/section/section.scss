@use "../../design/layout.scss";
@use "../../design/colors.scss";
@use "../../design/list.scss";
@use "sass:map";

$sizes: (
  sm: 1rem,
  md: 1.25rem,
  lg: 1.5rem,
);

$heights: (
  sm: 30px,
  md: 35px,
  lg: 40px,
);

.rc-section {
  @extend %col;
  margin: 2rem 0;
  width: 100%;
}

.rc-section-header {
  border-bottom: 1px solid colors.$secondary;
  color: colors.$black;
  font-size: 1rem;
  font-weight: 500;
  height: 30px;
  padding-left: 1rem;
  width: 100%;

  @each $key, $val in $sizes {
    &.rc-section-header-#{$key} {
      font-size: $val;
      height: map.get($heights, $key);
    }
  }
}

.rc-section-body {
  padding: 1rem;
  width: 100%;

  &:not(.rc-section-rtl) {
    @extend %left;
  }

  &.rc-section-rtl {
    @extend %right;
  }

  &:not(.rc-section-no-title) {
    min-height: calc(100% - 30px);
  }

  &.rc-section-no-title {
    min-height: 100%;
  }
}

.rc-section-header-rtl {
  @extend %right;
}
