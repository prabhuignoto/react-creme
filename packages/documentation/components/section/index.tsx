import { faSquare } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import React from 'react';
import DemoPageRenderer from '../../common/demo-page-renderer';
import Widgets from './widgets';

function index() {
  return (
    <DemoPageRenderer
      demoWidget={<Widgets />}
      title="Section"
      description="Section is a container for other components. It can be used to group components together."
      tabTitles={['Examples', 'Properties', 'Stackblitz']}
      stackBlitzCodes={['react-ts-2fdwsf']}
      pageIcon={<FontAwesomeIcon icon={faSquare} size="2x" />}
      sourceId="section/section.tsx"
      editId="section"
      properties={[
        {
          default: '',
          description: 'Title of the section',
          name: 'title',
          optional: 'Yes',
          type: 'String',
        },
        {
          default: '150',
          description: 'Minimum height of the section',
          name: 'height',
          optional: 'Yes',
          type: 'Number',
        },
        {
          default: 'row',
          description:
            'Layout of the section. can be <em>row</em> or <em>column</em>',
          name: 'layout',
          optional: 'Yes',
          type: 'String',
        },
        {
          default: 'False',
          description: 'Right to Left',
          name: 'RTL',
          optional: 'Yes',
          type: 'Boolean',
        },
        {
          default: 'sm',
          description:
            'Size of the page header. can be <em>sm</em> or <em>md</em> or <em>lg</em>.',
          name: 'size',
          optional: 'Yes',
          type: 'String',
        },
      ]}
    ></DemoPageRenderer>
  );
}

export default index;
