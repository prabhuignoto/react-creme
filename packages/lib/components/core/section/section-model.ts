import React from 'react';

export interface SectionProps {
  RTL?: boolean;
  border?: boolean;
  children: React.ReactNode | React.ReactNode[];
  hashPrefix?: string;
  height?: number;
  layout?: 'row' | 'column';
  noPadding?: boolean;
  size?: 'sm' | 'md' | 'lg';
  title?: string;
  useHash?: boolean;
}
