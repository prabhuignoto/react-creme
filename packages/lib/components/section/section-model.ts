import React from "react";

export interface SectionModel {
  children: React.ReactNode | React.ReactNode[];
  title?: string;
  height?: number;
  layout?: "row" | "column";
  RTL?: boolean;
  size?: "sm" | "md" | "lg";
}
