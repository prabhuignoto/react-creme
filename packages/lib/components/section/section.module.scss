@use '../../design/core.scss';
@use '../../design/theme.scss';
@use '../../design/list.scss';
@use 'sass:map';

$sizes: (
  sm: 1rem,
  md: 1.25rem,
  lg: 1.5rem,
);

$heights: (
  sm: 30px,
  md: 40px,
  lg: 50px,
);

.section {
  @extend %col;
  margin: 2rem 0;
  padding: 0 0.5rem;
  width: calc(100% - 1rem);

  &.border {
    @extend %border;
  }
}

.header {
  @extend %left;
  font-weight: 500;
  padding-left: 1rem;
  text-decoration: none;
  width: 100%;

  &.dark {
    border-bottom: 1px solid rgba(theme.$secondary-rgb, 0.5);
    color: theme.$white;
  }

  &:not(.dark) {
    border-bottom: 1px solid rgba(theme.$secondary-rgb, 0.5);
    color: theme.$primary;
  }

  @each $key, $val in $sizes {
    &.header_#{$key} {
      @extend %font-#{$key};
      height: map.get($heights, $key);
    }
  }
}

.body {
  width: calc(100% - 2rem);

  &.no_padding {
    padding: 0;
  }

  &:not(.no_padding) {
    padding: 1rem;
  }

  &:not(.rtl) {
    @extend %left;
  }

  &.rtl {
    direction: rtl;
  }

  @each $key, $val in $sizes {
    &.body_#{$key}:not(.no_title) {
      // height: map.get($heights, $key);
      min-height: calc(100% - map.get($heights, $key));
    }
  }

  // &:not(.no_title) {
  // }

  &.no_title {
    min-height: 100%;
  }
}

.header_rtl {
  @extend %right;
}
