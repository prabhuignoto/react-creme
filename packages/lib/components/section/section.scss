@use '../../design/core.scss';
@use '../../design/theme.scss';
@use '../../design/list.scss';
@use 'sass:map';

$sizes: (
  sm: 1rem,
  md: 1.25rem,
  lg: 1.5rem,
);

$heights: (
  sm: 30px,
  md: 35px,
  lg: 40px,
);

.rc-section {
  @extend %col;
  margin: 2rem 0;
  width: 100%;
}

.rc-section-header {
  @extend %left;
  border-bottom: 1px solid theme.$secondary;
  color: theme.$primary;
  font-size: 1rem;
  font-weight: 500;
  height: 30px;
  padding-left: 1rem;
  text-decoration: none;
  width: 100%;

  @each $key, $val in $sizes {
    &.rc-section-header-#{$key} {
      @extend %font-#{$key};
      height: map.get($heights, $key);
    }
  }
}

.rc-section-body {
  width: calc(100% - 2rem);

  &.rc-section-no-padding {
    padding: 0;
  }

  &:not(.rc-section-no-padding) {
    padding: 1.5rem 1rem 1rem;
  }

  &:not(.rc-section-rtl) {
    @extend %left;
  }

  &.rc-section-rtl {
    // @extend %right;
    direction: rtl;
  }

  &:not(.rc-section-no-title) {
    min-height: calc(100% - 30px);
  }

  &.rc-section-no-title {
    min-height: 100%;
  }
}

.rc-section-header-rtl {
  @extend %right;
}
