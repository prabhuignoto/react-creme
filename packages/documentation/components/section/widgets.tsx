import { Section } from '../../../lib/components';
import { DemoWidget } from '../../common/demo-widget';

function Widgets() {
  return (
    <div className="rc-demo-widgets">
      <DemoWidget name="Section" fullWidth>
        <Section size="md" title="Default rendering">
          <p>lorem ipsum </p>
        </Section>
      </DemoWidget>
      <DemoWidget name="Section" fullWidth>
        <Section size="md" title="Right to Left" RTL>
          <p>lorem ipsum </p>
        </Section>
      </DemoWidget>
    </div>
  );
}

export default Widgets;
