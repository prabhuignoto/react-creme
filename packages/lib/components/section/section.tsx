import classNames from 'classnames';
import React, { CSSProperties, useMemo } from 'react';
import { SectionProps } from './section-model';
import './section.scss';

const Section: React.FC<SectionProps> = ({
  children,
  title,
  height = 150,
  layout = 'row',
  RTL = false,
  size = 'sm',
  noPadding = false,
  useHash = false,
}) => {
  const sectionStyle = useMemo(
    () =>
      ({
        minHeight: `${height}px`,
      } as CSSProperties),
    [height]
  );

  const bodyStyle = useMemo(
    () =>
      ({
        alignItems: layout === 'column' ? 'center' : 'stretch',
        flexDirection: layout === 'column' ? 'row' : 'column',
      } as CSSProperties),
    []
  );

  const bodyClass = useMemo(
    () =>
      classNames('rc-section-body', {
        [`rc-section-${layout}`]: true,
        'rc-section-no-padding': noPadding,
        'rc-section-no-title': !title,
        'rc-section-rtl': RTL,
      }),
    [layout, title]
  );

  const headerClass = useMemo(() => {
    return classNames('rc-section-header', {
      'rc-section-header-rtl': RTL,
      [`rc-section-header-${size}`]: true,
    });
  }, []);

  const getTitle = useMemo(() => {
    return useHash ? `# ${title}` : title;
  }, []);

  const getId = useMemo(() => {
    return `${title?.toLowerCase().replace(' ', '-')}`;
  }, []);

  return (
    <div style={sectionStyle} className={'rc-section'}>
      {getTitle && (
        <div className={headerClass} id={getId}>
          {getTitle}
        </div>
      )}
      <div style={bodyStyle} className={bodyClass}>
        {children}
      </div>
    </div>
  );
};

Section.displayName = 'Section';

export { Section };
