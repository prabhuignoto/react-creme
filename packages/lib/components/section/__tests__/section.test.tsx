import { render } from '@testing-library/react';
import { Section } from '../section';

describe('Section', () => {
  it('should render Section correctly', () => {
    const { container, getByText } = render(
      <Section title="test">
        <span></span>
      </Section>
    );
    expect(container.firstChild).toMatchSnapshot();
    expect(getByText('test')).toBeInTheDocument();
  });

  it('should render section content', () => {
    const { getByText } = render(<Section>Section content</Section>);
    expect(getByText('Section content')).toBeInTheDocument();
  });

  it('should render title with hash', () => {
    const { getByText } = render(
      <Section useHash title="test">
        content
      </Section>
    );
    expect(getByText('# test')).toBeInTheDocument();
  });
});
