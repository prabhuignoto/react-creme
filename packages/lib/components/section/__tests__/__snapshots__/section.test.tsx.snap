// Vitest Snapshot v1

exports[`Section > should render Section correctly 1`] = `
<div
  class="_section_c93pb_1"
  role=""
  style="min-height: 150px;"
>
  <div
    class="_section_header_c93pb_1 _section_header_sm_c93pb_1"
    id="test"
    role="heading"
  >
    test
  </div>
  <div
    class="_section_body_c93pb_13 undefined"
    style="align-items: stretch; flex-direction: column;"
  >
    <span />
  </div>
</div>
`;
